src/vos/tests/vts_mvcc.c

```diff
@@ -0,0 +1,656 @@
+	print_message("  %s(%s, "DF_U64") (expect %d): ",
+		      r->o_name, rp, re, expected_rrc);
+	rc = r->o_func(arg, rp, re);
+	print_message("%d\n", rc);
+	if (rc != expected_rrc) {
+		nfailed++;
+		goto out;
+	}
+
+	if (re >= we)
+		expected_wrc = -DER_AGAIN;
+	if (is_rw(w)) {
+		bool e;
+
+		/*
+		 * What shall be w's read result? This shall override its write
+		 * result.
+		 */
+		if (re > we) {
+			e = true;
+		} else {
+			e = empty;
+			if (expected_rrc == 0 && is_rw(r))
+				e = (r->o_wtype == W_E);
+		}
+		if (w->o_rtype == R_E && !e)
+			expected_wrc = -DER_EXIST;
+		else if (w->o_rtype == R_NE && e)
+			expected_wrc = -DER_NONEXIST;
+	}
+	print_message("  %s(%s, "DF_U64") (expect %d): ",
+		      w->o_name, wp, we, expected_wrc);
+	rc = w->o_func(arg, wp, we);
+	print_message("%d\n", rc);
+	if (rc != expected_wrc)
+		nfailed++;
+
+out:
+	if (nfailed > 0)
+		print_message("FAILED: CASE %d.%d: %s, %s(%s, "DF_U64
+			      "), %s(%s, "DF_U64") [%d]\n", i, j,
+			      empty ? "empty" : "nonemtpy", r->o_name, rp, re,
+			      w->o_name, wp, we, mvcc_arg->i);
+	return nfailed;
+}
+
+/* Return the number of failures observed. */
+static int
+conflicting_rw_exec(struct io_test_args *arg, int i, struct op *r, struct op *w)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	daos_epoch_t	 re;			/* r epoch */
+	daos_epoch_t	 we;			/* w epoch */
+	char		 rp[L_COUNT + 1];	/* r path */
+	char		 wp[L_COUNT + 1];	/* w path */
+	char		*path_template = "coda";
+	bool		 emptiness[] = {true, false};
+	int		 j = 0;
+	int		 k;
+	int		 nfailed = 0;
+
+	/* T_R operations do not leave read epoch records at the moment. */
+	if (is_r(r))
+		return nfailed;
+
+	/* Set overlapping paths. */
+	set_path(r, path_template, rp);
+	set_path(w, path_template, wp);
+	D_ASSERTF(overlap(rp, wp), "overlap(\"%s\", \"%s\")", rp, wp);
+
+	for (k = 0; k < ARRAY_SIZE(emptiness); k++) {
+		bool empty = emptiness[k];
+
+		/* Read, then write. re > we. */
+		re = mvcc_arg->epoch + 10;
+		we = mvcc_arg->epoch;
+		nfailed += conflicting_rw_exec_one(arg, i, j, empty, r, rp,
+						     re, w, wp, we);
+		j++;
+		mvcc_arg->i++;
+		mvcc_arg->epoch += 100;
+
+		/* Read, then write. re == we. */
+		re = mvcc_arg->epoch;
+		we = mvcc_arg->epoch;
+		nfailed += conflicting_rw_exec_one(arg, i, j, empty, r, rp,
+						     re, w, wp, we);
+		j++;
+		mvcc_arg->i++;
+		mvcc_arg->epoch += 100;
+
+		/* Read, then write. re < we. Shall write OK. */
+		re = mvcc_arg->epoch;
+		we = mvcc_arg->epoch + 10;
+		nfailed += conflicting_rw_exec_one(arg, i, j, empty, r, rp,
+						     re, w, wp, we);
+		j++;
+		mvcc_arg->i++;
+		mvcc_arg->epoch += 100;
+	}
+
+	return nfailed;
+}
+
+/* Verify that a read causes a conflicting write to be rejected. */
+static void
+conflicting_rw(void **state)
+{
+	struct io_test_args	*arg = *state;
+	struct op		*r;
+	struct op		*w;
+	int			 n = 0;
+	int			 nfailed = 0;
+
+	/* For each readwrite... */
+	for_each_op(r) {
+		if (!(is_r(r) || is_rw(r)))
+			continue;
```

Comment:
Yes, to both questions. But I will keep the checks as they are, since I intentionally wrote them this way.

Reply:
At line 547, "if (is_r(r)) return nfailed;" means that we skip the test for "T_R" type. So here, we only need to handle "is_rw(r)"? that will reduce internal test cycles.

Reply:
I understand and am aware of this duplication. Here the condition is meant to match the comment two lines above; the check on line 547 is temporarily because reads' read timestamp updates have not implemented yet.